samples/outreach/blogs/serving_blogpost/data_utils.py

```diff
@@ -0,0 +1,204 @@
+  for i, url in enumerate(urls):
+    print('Downloading: {}'.format(url))
+    get_single_zip_data(url, output_file, i == 0)
+
+  output_file.close()
+
+
+def get_single_zip_data(url, output_file, first=False):
+  url = urllib.request.urlopen(url)
+  with ZipFile(BytesIO(url.read())) as my_zip_file:
+    contained_file = my_zip_file.namelist()[0]
+    for i, line in enumerate(my_zip_file.open(contained_file).readlines()):
+      if i == 0 and not first or '-9999' in str(line):
+        # Skip in the case of a header (and not the first file) or invalid data
+        continue
+      output_file.write(line)
+
+class Dataset():
+  def __init__(self,
+               data_path='./data/merge_2006_2012.csv',
+               batch_size=128,
+               data_points_day=144,
+               num_days=10,
+               steps=6,
+               mode='train'):
+    self.num_days = num_days
+    self.data_points_day = data_points_day
+    self.data_path = data_path
+    # Look back 10 days
+    self.lookback = data_points_day * num_days
+    # observations once every hour
+    self.steps = steps
+    # We will target one day in the future (24 hours)
+    self.delay = data_points_day
+    self.batch_size = batch_size
+    if mode == 'inference':
+      print('Setting batch size to 1 in inference mode')
+      self.batch_size = 1
+
+  def _preprocess_dataset_element(self, idx, data, lookback, steps, delay):
+    """fn to return the actual data given a specific index"""
+    # We map this function onto individual elements of the dataset
+    sample = data[idx - lookback: idx: steps]
+    target = tf.expand_dims(data[idx + delay][1], -1)
+    return sample, target
+
+  def get_dataset(self,
+                  data,
+                  shuffle=False):
+    """
+    Returns a tf.data.Dataset object containing the data associated with the
+    jena temperature data
+
+    Arguments:
+      data:  raw normalized data
+      shuffle:  boolean true false for shuffling the data or not
+
+    Returns:
+      samples of size (batch, lookback // steps, num features),
+             targets of size (batch,)
+    """
+    num_x = len(data)
+    idxs = np.arange(self.lookback, num_x - self.delay)
+    dataset = tf.data.Dataset.from_tensor_slices((idxs))
+    cfg = {
+      'data': tf.constant(data, dtype=tf.float32),
+      'lookback': self.lookback,
+      'steps': self.steps,
+      'delay': self.delay,
+    }
+    preproc_fn = functools.partial(self._preprocess_dataset_element, **cfg)
+
+    # Map our preprocessing function to every element in our dataset, taking
+    # advantage of multithreading
+    dataset = dataset.map(preproc_fn, num_parallel_calls=5)
+    if shuffle:
+      dataset = dataset.shuffle(num_x)
+    # It's necessary to repeat our data for all epochs
+    dataset = dataset.repeat().batch(self.batch_size).prefetch(1)
+    return dataset
+
+  def get_raw(self):
+    """
+    Loads the raw csv and returns the data within the csv without the header
```

Comment:
Last I recall, I had issues with loading with `pandas` since the csv file isn't `utf-8` encoded (I'm not actually sure what encoding it is)? I didn't look into it much further since working natively worked fine